// markdowntorender/src/index.js
/**
 * Content Renderer package
 * A collection of utilities for rendering various content formats
 */

const katexRenderer = require('./katex-renderer');
const mermaidRenderer = require('./mermaid-renderer');
const syntaxHighlighter = require('./syntax-highlighter');
const markdownRenderer = require('./markdown-renderer');

module.exports = {
  // KaTeX math rendering
  getKaTexResources: katexRenderer.getKaTexResources,
  processMathFormulas: katexRenderer.processMathFormulas,
  
  // Mermaid diagram rendering
  getMermaidResources: mermaidRenderer.getMermaidResources,
  processMermaidDiagrams: mermaidRenderer.processMermaidDiagrams,
  
  // Syntax highlighting
  getHighlightJsResources: syntaxHighlighter.getHighlightJsResources,
  processCodeBlocks: syntaxHighlighter.processCodeBlocks,
  
  // Markdown rendering
  createMarkdownRenderer: markdownRenderer.createMarkdownRenderer,
  renderMarkdown: markdownRenderer.renderMarkdown,
  renderMarkdownToHtml: markdownRenderer.renderMarkdownToHtml,
  renderMarkdownFile: markdownRenderer.renderMarkdownFile
};
