// markdowntorender/src/katex-renderer.js
/**
 * Module for rendering math formulas using KaTeX
 */

/**
 * Clean math expressions to avoid common issues
 * @param {string} expr - The math expression to clean
 * @returns {string} - Cleaned math expression
 */
function cleanMathExpression(expr) {
  return expr
    .replace(/\\<br>/g, '\\\\')
    .replace(/\\</g, '\\\\')
    .replace(/<br>/g, '\\\\');
}

/**
 * Get KaTeX resources (CSS and script tags)
 * @param {Object} options - Configuration options
 * @param {string} options.version - KaTeX version to use (default: '0.16.9')
 * @returns {Object} - Object with CSS and script properties
 */
function getKaTexResources(options = {}) {
  const version = options.version || '0.16.9';
  
  return {
    css: `<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@${version}/dist/katex.min.css">`,
    script: `<script src="https://cdn.jsdelivr.net/npm/katex@${version}/dist/katex.min.js"></script>`
  };
}

/**
 * Process math formulas in HTML content
 * @param {string} html - HTML content with math formulas
 * @param {Object} options - Configuration options
 * @param {boolean} options.displayMode - Display mode for math rendering (default: false)
 * @param {boolean} options.autoRender - Whether to auto-render formulas (default: true)
 * @param {Object} options.configOverrides - Additional KaTeX configuration options
 * @returns {string} - Processed HTML with KaTeX initialization script
 */
function processMathFormulas(html, options = {}) {
  const displayMode = options.displayMode !== undefined ? options.displayMode : false;
  const autoRender = options.autoRender !== undefined ? options.autoRender : true;
  const configOverrides = options.configOverrides || {};
  
  // If auto-render is disabled, return the HTML as is
  if (!autoRender) return html;
  
  // Add auto-render script for KaTeX
  const katexScript = `
    <script src="https://cdn.jsdelivr.net/npm/katex@0.16.9/dist/contrib/auto-render.min.js"></script>
    <script>
      document.addEventListener("DOMContentLoaded", function() {
        renderMathInElement(document.body, {
          delimiters: [
            {left: "$$", right: "$$", display: true},
            {left: "$", right: "$", display: false}
          ],
          throwOnError: false,
          displayMode: ${displayMode},
          ...${JSON.stringify(configOverrides)}
        });
      });
    </script>
  `;
  
  return html + katexScript;
}

module.exports = {
  getKaTexResources,
  processMathFormulas,
  cleanMathExpression
};
